src/tiered/tiered_handle.c

```diff
@@ -228,6 +228,58 @@ __tiered_create_local(WT_SESSION_IMPL *session, WT_TIERED *tiered)
     return (ret);
 }
 
+/*
+ * __tiered_flush_older_objects --
+ *     Check earlier objects and see if flush work units for earlier objects need to be pushed onto
+ *     the work queue. This can happen if the system crashed during a flush tier.
+ */
+static int
+__tiered_flush_older_objects(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+    WT_CONFIG_ITEM cval;
+    WT_DECL_RET;
+    uint32_t i;
+    const char *obj_name, *obj_uri, *obj_val;
+    bool exist;
+
+    /* If this is a new table and there are no objects there is nothing to do. */
+    if (tiered->current_id == 0)
+        return (0);
+
+    /*
+     * Work our way backwards through all earlier objects and look at each object's flush_time
+     * configuration setting. If it is zero that means the object is not flushed and a work unit
+     * will be pushed for that object.
+     */
+    obj_name = obj_uri = obj_val = NULL;
+    exist = false;
+    for (i = tiered->current_id - 1; i > 0; --i) {
+        WT_ERR(__wt_tiered_name(session, &tiered->iface, i, WT_TIERED_NAME_OBJECT, &obj_uri));
+        obj_name = obj_uri;
+        WT_PREFIX_SKIP_REQUIRED(session, obj_name, "object:");
+        WT_ERR(__wt_fs_exist(session, obj_name, &exist));
+        /*
+         * We only need to worry about objects that exist locally. As soon as we find an object that
+         * does not exist, that means that it, and all earlier objects, have been flushed.
+         */
+        if (exist) {
+            WT_ERR(__wt_metadata_search(session, obj_uri, (char **)&obj_val));
+            WT_ERR(__wt_config_getones(session, obj_val, "flush_time", &cval));
+            __wt_verbose(session, WT_VERB_TIERED,
+              "OLDER_OBJECTS: local object %s has flush time %d", obj_uri, (int)cval.val);
+            if (cval.val == 0)
+                WT_ERR(__wt_tiered_put_flush(session, tiered, i));
```

Comment:
On first looking at this I was puzzled why we can't stop looking for things to flush when we hit on with `flush_time` greater than zero.  I realized that we need to keep going because if there are N objects to flush, and we fail again during flushing, we have no guarantees are about order in which those flushes might complete.  So we have to check every local file.  

Assuming I've understood this correctly, it might be worth a comment to explain that.  Other engineers might be as easily confused as me :-)

Reply:
Sure. I had to reason about that too. I will add a comment.